src/crimson/os/seastore/segment_cleaner.cc

```diff
@@ -590,130 +634,263 @@ SegmentCleaner::gc_cycle_ret SegmentCleaner::do_gc_cycle()
+      [this, &t](auto &limit) {
+      return trim_backrefs(t, limit).si_then(
+	[this, &t, &limit](auto trim_backrefs_to)
+	-> ExtentCallbackInterface::submit_transaction_direct_iertr::future<
+	     journal_seq_t> {
+	if (trim_backrefs_to != JOURNAL_SEQ_NULL) {
+	  return ecb->submit_transaction_direct(
+	    t, std::make_optional<journal_seq_t>(trim_backrefs_to)
+	  ).si_then([trim_backrefs_to=std::move(trim_backrefs_to)]() mutable {
+	    return seastar::make_ready_future<
+	      journal_seq_t>(std::move(trim_backrefs_to));
+	  });
+	}
+	return seastar::make_ready_future<journal_seq_t>(std::move(limit));
+      });
+    });
+  }).handle_error(
+    crimson::ct_error::eagain::handle([](auto) {
+      ceph_abort("unexpected eagain");
+    }),
+    crimson::ct_error::pass_further_all()
+  ).safe_then([this](auto seq) {
+    return repeat_eagain([this, seq=std::move(seq)]() mutable {
+      return ecb->with_transaction_intr(
+	Transaction::src_t::CLEANER_TRIM,
+	"trim_journal",
+	[this, seq=std::move(seq)](auto& t)
+      {
+	return rewrite_dirty(t, seq
+	).si_then([this, &t] {
+	  return ecb->submit_transaction_direct(t);
+	});
       });
     });
   });
 }
 
+SegmentCleaner::retrieve_backref_extents_ret
+SegmentCleaner::_retrieve_backref_extents(
+  Transaction &t,
+  std::set<
+    Cache::backref_extent_buf_entry_t,
+    Cache::backref_extent_buf_entry_t::cmp_t> &&backref_extents,
+  std::vector<CachedExtentRef> &extents)
+{
+  return trans_intr::parallel_for_each(
+    backref_extents,
+    [this, &extents, &t](auto &ent) {
+    // only the gc fiber which is single can rewrite backref extents,
+    // so it must be alive
+    assert(is_backref_node(ent.type));
+    LOG_PREFIX(SegmentCleaner::_retrieve_backref_extents);
+    DEBUGT("getting backref extent of type {} at {}",
+      t,
+      ent.type,
+      ent.paddr);
+    return cache.get_extent_by_type(
+      t, ent.type, ent.paddr, L_ADDR_NULL, BACKREF_NODE_SIZE
+    ).si_then([&extents](auto ext) {
+      extents.emplace_back(std::move(ext));
+    });
+  });
+}
+
+SegmentCleaner::retrieve_live_extents_ret
+SegmentCleaner::_retrieve_live_extents(
+  Transaction &t,
+  std::set<
+    backref_buf_entry_t,
+    backref_buf_entry_t::cmp_t> &&backrefs,
+  std::vector<CachedExtentRef> &extents)
+{
+  return seastar::do_with(
+    JOURNAL_SEQ_NULL,
+    std::move(backrefs),
+    [this, &t, &extents](auto &seq, auto &backrefs) {
+    return trans_intr::do_for_each(
+      backrefs,
+      [this, &extents, &t, &seq](auto &ent) {
+      LOG_PREFIX(SegmentCleaner::gc_reclaim_space);
+      DEBUGT("getting extent of type {} at {}~{}",
+	t,
+	ent.type,
+	ent.paddr,
+	ent.len);
+      return ecb->get_extent_if_live(
+	t, ent.type, ent.paddr, ent.laddr, ent.len
+      ).si_then([this, &extents, &ent, &seq](auto ext) {
+	if (!ext) {
+	  logger().debug(
+	    "SegmentCleaner::gc_reclaim_space:"
+	    " addr {} dead, skipping",
+	    ent.paddr);
+	  auto backref = cache.get_del_backref(ent.paddr);
```

Comment:
Why can this still happen? As far as I understand, we should have already coverred and applied this paddr range with `cache.get_(del_)backrefs_in_range()` in a previous step, right?

Reply:
Um, not really, chances are that, after we calculate the set of live extents and before we successfully retrieve those extents, some of them may be removed/overwritten by io transactions, we need to take care of that scenario.

Reply:
Ah, right. Cache is kept updated by other transactions at the same time.